### PR TITLE
ONNX runtime support

### PR DESCRIPTION
**Pull request by karl-joan**
Made changes to pull request #904 since the author didn't respond.

Changes:
- Separated functions for image preprocessing and post-processing.
- You can supply options to `onnxruntime`.
- Fixed bugs in image post-processing, where `iou_threshold` was missing and bounding boxes were sent to `non_max_supression` in wrong format.
- Fixed bug where sessions options weren't passed to `onnxruntime`.
- Added some test.
- Renamed `onnx.py` to `onnx_model.py` to avoid confusion with `onnx` library.

The code was designed with the intention of being used for YOLOv8 ONNX model. Perhaps some changes should be made to make it more suitable for other architectures.

### CODE REVIEW

**Comment by karl-joan** [comment]
@julled You can go ahead and test it

**Comment by LongLe18** [comment]
@karl-joan I tested your contribute. It's appear error like image below. And image which I used to test (https://ultralytics.com/images/bus.jpg)
image

**Comment by karl-joan** [comment]
@LongLe18 The test is meant to be run with image specified in the test file. If you change the image, then it won't work since the image you provided does not contain any cars.

If you just want to try the model, then use one of the built in functions such as `get_sliced_prediction` or `get_prediction`.

**Comment by LongLe18** [comment]
@karl-joan  ok tks sir. I understood, you have not handled except yet. I will try again

**Comment by fcakyon** [comment]
Thanks a lot for this amazing contribution @karl-joan !

If you could add a demo notebook as other model implementations, we would gladly merge the PR 👍

**Review by fcakyon — CHANGES_REQUESTED** [review]
code styling errors need to be fixed, a demo notebook need to be added

File: requirements.txt [hunk 1]
```diff
@@ -8,3 +8,4 @@ fire
 terminaltables
 requests
 click
+onnxruntime
```
Comment (fcakyon) [inline requirements.txt:11]: it should not be a core dependency

```suggestion
```

File: sahi/auto_model.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
     "torchvision": "TorchVisionDetectionModel",
     "yolov5sparse": "Yolov5SparseDetectionModel",
     "yolonas": "YoloNasDetectionModel",
+    "onnx_model": "ONNXDetectionModel"
```
Comment (fcakyon) [inline sahi/auto_model.py:14]:
```suggestion
    "yolov8onnx": "Yolov8OnnxDetectionModel"
```

File: sahi/models/__init__.py [hunk 3]
```diff
@@ -1 +1 @@
-from . import base, detectron2, huggingface, mmdet, torchvision, yolonas, yolov5
+from . import base, detectron2, huggingface, mmdet, onnx_model, torchvision, yolonas, yolov5
```
Comment (fcakyon) [inline sahi/models/__init__.py:1]:
```suggestion
from . import base, detectron2, huggingface, mmdet, yolov8onnx, torchvision, yolonas, yolov5
```

File: sahi/models/onnx_model.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+logger = logging.getLogger(__name__)
+
+from sahi.models.base import DetectionModel
+from sahi.prediction import ObjectPrediction
+from sahi.utils.compatibility import fix_full_shape_list, fix_shift_amount_list
+from sahi.utils.import_utils import check_requirements
+from sahi.utils.onnx_model import non_max_supression, xywh2xyxy
+
+
+class ONNXDetectionModel(DetectionModel):
```
Comment (fcakyon) [inline sahi/models/onnx_model.py:20]:
```suggestion
class Yolov8OnnxDetectionModel(DetectionModel):
```

File: tests/test_onnxmodel.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+from sahi.utils.onnx_model import Yolov8ONNXTestConstants, download_yolov8n_onnx_model
+
+MODEL_DEVICE = "cpu"
+CONFIDENCE_THRESHOLD = 0.3
+IOU_THRESHOLD = 0.7
+IMAGE_SIZE = 640
+
+
+class TestONNXDetectionModel(unittest.TestCase):
```
Comment (fcakyon) [inline tests/test_onnxmodel.py:17]:
```suggestion
class TestYolov8OnnxDetectionModel(unittest.TestCase):
```

File: tests/test_onnxmodel.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    def test_load_model(self):
+        from sahi.models.onnx_model import ONNXDetectionModel
```
Comment (fcakyon) [inline tests/test_onnxmodel.py:19]:
```suggestion
        from sahi.models.yolov8onnx import Yolov8OnnxDetectionModel
```

**Comment by fcakyon** [comment]
@karl-joan can you please fix the styling? https://github.com/obss/sahi#contributing

**Comment by karl-joan** [comment]
@fcakyon  Yes, I'll try to get that and the notebook tomorrow done.

**Comment by karl-joan** [comment]
@fcakyon Let me know if you need anything else.

**Comment by devrimcavusoglu** [comment]
@karl-joan @fcakyon what is the status here ? I can take the lead, but I assume code formatting/linting has been addressed. As far as I reviewed the files, there are no additional requirements.

**Comment by karl-joan** [comment]
@devrimcavusoglu It's currently waiting behind @fcakyon

**Review by devrimcavusoglu — APPROVED** [review]
LGTM!

**Pull request merged**